platforms/chibios/drivers/vendor/RP/RP2040/ps2_vendor.c

```diff
@@ -0,0 +1,335 @@
+// Copyright 2022 Marek Kraus (@gamelaster)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "quantum.h"
+
+#include "hardware/pio.h"
+#include "hardware/clocks.h"
+#include "ps2.h"
+#include "print.h"
+
+#if !defined(MCU_RP)
+#    error PIO Driver is only available for Raspberry Pi 2040 MCUs!
+#endif
+
+#if defined(PS2_ENABLE)
+#    if defined(PS2_MOUSE_ENABLE)
+#        if !defined(PS2_MOUSE_USE_REMOTE_MODE)
+#            define BUFFERED_MODE_ENABLE
+#        endif
+#    else // PS2 Keyboard
+#        define BUFFERED_MODE_ENABLE
+#    endif
+#endif
+
+#if PS2_DATA_PIN + 1 != PS2_CLOCK_PIN
+#    error PS/2 Clock pin must be followed by data pin!
+#endif
+
+static inline void pio_serve_interrupt(void);
+
+#if defined(PS2_PIO_USE_PIO1)
+static const PIO pio = pio1;
+
+OSAL_IRQ_HANDLER(RP_PIO1_IRQ_0_HANDLER) {
+    OSAL_IRQ_PROLOGUE();
+    pio_serve_interrupt();
+    OSAL_IRQ_EPILOGUE();
+}
+#else
+static const PIO pio = pio0;
+
+OSAL_IRQ_HANDLER(RP_PIO0_IRQ_0_HANDLER) {
+    OSAL_IRQ_PROLOGUE();
+    pio_serve_interrupt();
+    OSAL_IRQ_EPILOGUE();
+}
+#endif
+
+#define PS2_WRAP_TARGET 2
+#define PS2_WRAP 30
+
+// clang-format off
+static const uint16_t ps2_program_instructions[] = {
+    0xe080, //  0: set    pindirs, 0
+    0xe000, //  1: set    pins, 0
```

Comment:
Can these instructions be moved into the init code?

Reply:
Yes! Resolved.